File: backend/internal/repository/device_test.go
package repository

import (
	"context"
	"regexp"
	"testing"
	"time"

	"github.com/DATA-DOG/go-sqlmock"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/chaojixinren/pulse/internal/model"
)

var deviceRepoCols = []string{"id", "user_id", "device_id", "name", "device_type", "firmware_version", "battery_level", "last_seen_at", "is_active", "device_token_hash", "created_at", "updated_at"}
var bindCodeRepoCols = []string{"id", "user_id", "code", "expires_at", "used_at", "created_at"}

func TestDeviceRepoCreate(t *testing.T) {
	db, mock := newMockDB(t)
	repo := NewDeviceRepo(db)

	mock.ExpectExec(regexp.QuoteMeta("INSERT INTO devices")).
		WithArgs("d1", "u1", "dev-1", "手表", "wearable", nil, nil, nil, true, "hash").
		WillReturnResult(sqlmock.NewResult(1, 1))

	err := repo.Create(context.Background(), &model.Device{ID: "d1", UserID: "u1", DeviceID: "dev-1", Name: "手表", DeviceType: "wearable", IsActive: true, DeviceTokenHash: "hash"})
	require.NoError(t, err)
	assert.NoError(t, mock.ExpectationsWereMet())
}

func TestDeviceRepoGetByDeviceID(t *testing.T) {
	db, mock := newMockDB(t)
	repo := NewDeviceRepo(db)
	now := time.Now().UTC()

	mock.ExpectQuery(regexp.QuoteMeta("SELECT id, user_id, device_id")).
		WithArgs("dev-1").
		WillReturnRows(sqlmock.NewRows(deviceRepoCols).AddRow("d1", "u1", "dev-1", "手表", "wearable", nil, nil, nil, true, "hash", now, now))

	d, err := repo.GetByDeviceID(context.Background(), "dev-1")
	require.NoError(t, err)
	require.NotNil(t, d)
	assert.Equal(t, "d1", d.ID)
	assert.NoError(t, mock.ExpectationsWereMet())
}

func TestDeviceRepoUpdateHeartbeat(t *testing.T) {
	db, mock := newMockDB(t)
	repo := NewDeviceRepo(db)
	firmware := "1.0"
	battery := 80

	mock.ExpectExec(regexp.QuoteMeta("UPDATE devices SET firmware_version")).
		WithArgs("1.0", 80, sqlmock.AnyArg(), sqlmock.AnyArg(), "d1", "u1").
		WillReturnResult(sqlmock.NewResult(0, 1))

	require.NoError(t, repo.UpdateHeartbeat(context.Background(), "d1", "u1", &firmware, &battery))
	assert.NoError(t, mock.ExpectationsWereMet())
}

func TestDeviceRepoCreateBindCodeAndMarkUsed(t *testing.T) {
	db, mock := newMockDB(t)
	repo := NewDeviceRepo(db)
	now := time.Now().UTC()

	mock.ExpectExec(regexp.QuoteMeta("INSERT INTO device_bind_codes")).
		WithArgs("c1", "u1", "123456", now).
		WillReturnResult(sqlmock.NewResult(1, 1))
	require.NoError(t, repo.CreateBindCode(context.Background(), &model.DeviceBindCode{ID: "c1", UserID: "u1", Code: "123456", ExpiresAt: now}))

	mock.ExpectExec(regexp.QuoteMeta("UPDATE device_bind_codes SET used_at")).
		WithArgs(sqlmock.AnyArg(), "123456").
		WillReturnResult(sqlmock.NewResult(0, 1))
	require.NoError(t, repo.MarkBindCodeUsed(context.Background(), "123456"))
	assert.NoError(t, mock.ExpectationsWereMet())
}

func TestDeviceRepoCreateCommand(t *testing.T) {
	db, mock := newMockDB(t)
	repo := NewDeviceRepo(db)

	mock.ExpectExec(regexp.QuoteMeta("INSERT INTO device_commands")).
		WithArgs("cmd1", "d1", "u1", "start_recording", "pending").
		WillReturnResult(sqlmock.NewResult(1, 1))

	require.NoError(t, repo.CreateCommand(context.Background(), &model.DeviceCommand{ID: "cmd1", DeviceID: "d1", UserID: "u1", Command: "start_recording", Status: "pending"}))
	assert.NoError(t, mock.ExpectationsWereMet())
}

File: frontend/src/utils/device.test.ts
import { describe, expect, it } from 'vitest';
import { COMMAND_LABELS, DEVICE_COMMANDS, formatBattery, isDeviceOnline } from './device';
import type { Device } from '@/types/device.types';

const base: Device = {
  id: 'd1',
  user_id: 'u1',
  device_id: 'hw-001',
  name: '我的设备',
  device_type: 'wearable',
  is_active: true,
  created_at: '',
  updated_at: '',
};

const NOW = new Date('2024-06-05T10:00:00Z').getTime();

describe('isDeviceOnline', () => {
  it('激活且最近心跳在阈值内视为在线', () => {
    const device: Device = { ...base, last_seen_at: '2024-06-05T09:58:00Z' };
    expect(isDeviceOnline(device, NOW)).toBe(true);
  });

  it('超过阈值视为离线', () => {
    const device: Device = { ...base, last_seen_at: '2024-06-05T09:00:00Z' };
    expect(isDeviceOnline(device, NOW)).toBe(false);
  });

  it('未激活视为离线', () => {
    const device: Device = { ...base, is_active: false, last_seen_at: '2024-06-05T09:58:00Z' };
    expect(isDeviceOnline(device, NOW)).toBe(false);
  });

  it('无心跳视为离线', () => {
    expect(isDeviceOnline(base, NOW)).toBe(false);
  });
});

describe('formatBattery', () => {
  it('有电量时展示百分比', () => {
    expect(formatBattery(80)).toBe('80%');
    expect(formatBattery(0)).toBe('0%');
  });

  it('未上报时展示占位符', () => {
    expect(formatBattery(undefined)).toBe('—');
  });
});

describe('DEVICE_COMMANDS / COMMAND_LABELS', () => {
  it('预设指令与后端允许的指令一致', () => {
    expect(DEVICE_COMMANDS.map((c) => c.value).sort()).toEqual([
      'start_recording',
      'stop_recording',
    ]);
    expect(COMMAND_LABELS.start_recording).toBe('开始录音');
    expect(COMMAND_LABELS.stop_recording).toBe('停止录音');
  });
});
